WordPress/gutenberg
# Autocomplete: refactor to TypeScript

## What?
Refactor `Autocomplete` component to TypeScript

Part of https://github.com/WordPress/gutenberg/issues/35744

## Why?

The refactor to TypeScript has many benefits (auto-generated docs, static linting and error prevention, better IDE experience). See https://github.com/WordPress/gutenberg/issues/35744 for more details

## How?
Followed the steps in the [TypeScript migration guide](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md?rgh-link-date=2022-11-29T14%3A45%3A34Z#refactoring-a-component-to-typescript)

## Testing Instructions
- Ensure all unit and e2e tests still pass
- In the editor, test an autocompleter like the block inserter or `@` mentions, confirm they work as expected

## Notes:
- There are a couple of items I've commented on the PR that I'd love specific feedback/advice on
- There are also a handful of runtime changes that felt needed. I've highlighted them with comments as well for feedback/alternate approaches. TIA 😄
